nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -275,6 +275,22 @@ class BtActionNode : public BT::ActionNodeBase
       }
     }
 
+    auto start_time = node_->now();
+    while (rclcpp::ok()
```

Comment:
I strongly agree with you, the while loop could be simplified. However:

- I think need to keep the timeout in order to prevent infinite looping/spinning

- I would rather not run a *`spin_some()` blindly* too, but we can not use the `future_cancel` to do so. Indeed the variable `future_cancel` can only be used to check on the **cancel** result availability. 
To check **goal** result availability, one can use the variable [`goal_result_available_`](https://github.com/wyca-robotics/navigation2/blob/04261fb7633c69baf709bae739310e37a3322812/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L280) (which is a boolean variable already implemented, not a future).  As this variable is only updated while the `callback_group_executor_` is spinning, the `spin_some()` has been added here to check on the **goal** result availability.

Reply:
My assertion (which you could test and let me know if I'm wrong) is that when the `future_cancel` returns true after spinning, that you're **guaranteed** to have a result message waiting in your callback queue for a `spin_some()` to pick up. If that is accurate, then there is no need for a `while()` loop at all, therefore no requirement for a timeout from that while loop.